dot/parachain/statement-distribution/known_backed_candidate.go

```diff
@@ -0,0 +1,223 @@
+// we have confirmed as having been backed.
+type KnownBackedCandidate struct {
+	groupIndex      parachaintypes.GroupIndex
+	localKnowledge  StatementFilter
+	mutualKnowledge map[parachaintypes.ValidatorIndex]MutualKnowledge
+}
+
+func (kbc *KnownBackedCandidate) hasReceivedManifestFrom(validator parachaintypes.ValidatorIndex) bool {
+	mk, ok := kbc.mutualKnowledge[validator]
+	if !ok {
+		return false
+	}
+
+	return mk.remoteKnowledge != nil
+}
+
+func (kbc *KnownBackedCandidate) hasSentManifestTo(validator parachaintypes.ValidatorIndex) bool {
+	mk, ok := kbc.mutualKnowledge[validator]
+	if !ok {
+		return false
+	}
+
+	return mk.localKnowledge != nil
+}
+
+func (kbc *KnownBackedCandidate) sentManifestTo(
+	validator parachaintypes.ValidatorIndex,
+	localKnowledge StatementFilter,
+) {
+	mk, ok := kbc.mutualKnowledge[validator]
+	if !ok {
+		mk = MutualKnowledge{
+			remoteKnowledge:   nil,
+			localKnowledge:    nil,
+			receivedKnowledge: nil,
+		}
+	}
+
+	groupSize := uint(localKnowledge.secondedInGroup.Len())
+	receivedKnowledge, err := NewStatementFilter(groupSize, false)
+	if err != nil {
+		logger.Warnf("failed to create statement filter instance with group size %d: %v", groupSize, err)
+	}
+
+	mk.receivedKnowledge = receivedKnowledge
+	mk.localKnowledge = &localKnowledge
+	kbc.mutualKnowledge[validator] = mk
+}
+
+func (kbc *KnownBackedCandidate) manifestReceivedFrom(
+	validator parachaintypes.ValidatorIndex,
+	remoteKnowledge StatementFilter,
+) {
+	mk, ok := kbc.mutualKnowledge[validator]
+	if !ok {
+		mk = MutualKnowledge{
+			remoteKnowledge:   nil,
+			localKnowledge:    nil,
+			receivedKnowledge: nil,
+		}
+	}
+
+	mk.remoteKnowledge = &remoteKnowledge
+	kbc.mutualKnowledge[validator] = mk
+}
+
+// directStatementSenders returns a map representing each potential sender(ValidatorIndex)
+// and if the sender should already know about the statement, because we just sent it to it.
+func (kbc *KnownBackedCandidate) directStatementSenders(
+	groupIndex parachaintypes.GroupIndex,
+	originatorIndexInGroup uint,
+	statementKind StatementKind,
+) map[parachaintypes.ValidatorIndex]bool {
+	senders := make(map[parachaintypes.ValidatorIndex]bool)
+
+	if groupIndex != kbc.groupIndex {
+		return senders
+	}
+
+	for validatorIndex, mk := range kbc.mutualKnowledge {
+		if mk.remoteKnowledge == nil {
+			continue
+		}
+
+		if mk.receivedKnowledge == nil || !mk.receivedKnowledge.Contains(originatorIndexInGroup, statementKind) {
+			continue
+		}
+
+		if mk.localKnowledge != nil && mk.localKnowledge.Contains(originatorIndexInGroup, statementKind) {
+			senders[validatorIndex] = true
+		}
+	}
+
+	return senders
+}
+
+func (kbc *KnownBackedCandidate) directStatementRecipients(
+	groupIndex parachaintypes.GroupIndex,
+	originatorIndexInGroup uint,
+	statementKind StatementKind,
+) []parachaintypes.ValidatorIndex {
+	recipients := make([]parachaintypes.ValidatorIndex, 0)
```

Comment:
let's also provide the capacity of slice to avoid potential memory reallocation.
```suggestion
recipients := make([]parachaintypes.ValidatorIndex, 0, len(kbc.mutualKnowledge))
```